File: app/src/jvx/java/art/coded/wireframe/view/custom/CustomView.java
package art.coded.wireframe.view.custom;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PointF;
import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.core.view.AccessibilityDelegateCompat;
import androidx.core.view.ViewCompat;
import androidx.core.view.accessibility.AccessibilityNodeInfoCompat;

import art.coded.wireframe.R;

/**
 * 
 */
public class CustomView extends View {

    private static final String LOG_TAG = CustomView.class.getSimpleName();

    private enum Level {
        OFF { @Override public final int label() { return R.string.level_off; } },
        LOW { @Override public final int label() { return R.string.level_low; } },
        MEDIUM { @Override public final int label() { return R.string.level_medium; } },
        HIGH { @Override public final int label() { return R.string.level_high; } };
        abstract public int label();
        public Level next() {
            switch(label()) {
                case R.string.level_off: return LOW;
                case R.string.level_low: return MEDIUM;
                case R.string.level_medium: return HIGH;
                default: return OFF;
            }
        }
    }

    private final int RADIUS_OFFSET_LABEL = 30;
    private final int RADIUS_OFFSET_INDICATOR = -35;

    private float mRadius = 0.0f;
    private final PointF mPointPosition = new PointF(0.0f, 0.0f);
    private Level mLevel = Level.OFF;

    private int levelLowColor = 0;
    private int levelMediumColor = 0;
    private int levelHighColor = 0;

    private Paint mPaint;

    public CustomView(Context context) {
        super(context);
        init(null, 0);
    }

    public CustomView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(attrs, 0);
    }

    public CustomView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(attrs, defStyle);
    }

    private void updateContentDescription() {
        setContentDescription(getResources().getString(mLevel.label()));
    }

    private void init(AttributeSet attrs, int defStyle) {
        setClickable(true);

        // Set up a default Paint object
        mPaint = new Paint();
        mPaint.setFlags(Paint.ANTI_ALIAS_FLAG);
        mPaint.setTextAlign(Paint.Align.CENTER);
        mPaint.setTextSize(55.0f);
        mPaint.setTypeface(Typeface.create("", Typeface.BOLD));

        ViewCompat.setAccessibilityDelegate(this, new AccessibilityDelegateCompat() {
            @Override public void onInitializeAccessibilityNodeInfo(
                    View host, AccessibilityNodeInfoCompat info) {
                super.onInitializeAccessibilityNodeInfo(host, info);
                AccessibilityNodeInfoCompat.AccessibilityActionCompat customClick =
                        new AccessibilityNodeInfoCompat.AccessibilityActionCompat(
                            AccessibilityNodeInfoCompat.ACTION_CLICK,
                            getContext().getString((mLevel !=  Level.HIGH) ? R.string.change : R.string.reset)
                        );
                info.addAction(customClick);
            }
        });

        TypedArray typedArray = getContext().obtainStyledAttributes(attrs, R.styleable.CustomView);
        levelLowColor = typedArray.getColor(R.styleable.CustomView_levelColor1, 0);
        levelMediumColor =  typedArray.getColor(R.styleable.CustomView_levelColor2, 0);
        levelHighColor =  typedArray.getColor(R.styleable.CustomView_levelColor3, 0);

        typedArray.recycle();

        updateContentDescription();
    }

    @Override public boolean performClick() {
        mLevel = mLevel.next();
        updateContentDescription();

        invalidate();
        super.performClick();
        return true;
    }

    @Override
    protected void onDraw(Canvas canvas) {

        switch(mLevel) {
            case OFF: mPaint.setColor(Color.GRAY);break;
            case LOW: mPaint.setColor(levelLowColor); break;
            case MEDIUM: mPaint.setColor(levelMediumColor); break;
            default: mPaint.setColor(levelHighColor);
        }

        canvas.drawCircle(getWidth()/ 2f, getHeight()/ 2f, mRadius, mPaint);

        final float markerRadius = mRadius + RADIUS_OFFSET_INDICATOR;
        computeXYForLevel(mLevel, markerRadius);
        mPaint.setColor(Color.BLACK);
        canvas.drawCircle(mPointPosition.x, mPointPosition.y, mRadius/12, mPaint);

        final float labelRadius = mRadius + RADIUS_OFFSET_LABEL;
        for (Level i : Level.values()) {
            computeXYForLevel(i, labelRadius);
            final String label = getResources().getString(i.label());
            canvas.drawText(label, mPointPosition.x, mPointPosition.y, mPaint);
        }
    }

    @Override
    public void setAccessibilityDelegate(@Nullable AccessibilityDelegate delegate) {
        super.setAccessibilityDelegate(delegate);
    }

    @Override protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        mRadius = Math.min(w, h) / 2.0f * 0.8f;
    }

    private void computeXYForLevel(Level pos, Float radius) {
        // Angles are in radians.
        final double startAngle = Math.PI * (9 / 8.0);
        final double angle = startAngle + pos.ordinal() * (Math.PI / 4);
        mPointPosition.x = ((float) (radius * Math.cos(angle))) + getWidth() / 2f;
        mPointPosition.y = ((float) (radius * Math.sin(angle))) + getHeight() / 2f;
    }

    public int getLevelLabelRes() { return mLevel.label(); }
}

File: app/src/ktx/java/art/coded/wireframe/view/custom/CustomView.kt
package art.coded.wireframe.view.custom

import android.content.Context
import android.graphics.*
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.view.View
import art.coded.wireframe.R
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.sin

private enum class Level(val label : Int) {
    OFF(R.string.level_off),
    LOW(R.string.level_low),
    MEDIUM(R.string.level_medium),
    HIGH(R.string.level_high);

    fun next() = when (this) {
        OFF -> LOW
        LOW -> MEDIUM
        MEDIUM -> HIGH
        HIGH -> OFF
    }
}

private const val RADIUS_OFFSET_LABEL = 30
private const val RADIUS_OFFSET_INDICATOR = -35

/**
 * TODO: document your custom view class.
 */
class CustomView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0
) : View(context, attrs, defStyle) {

    private var radius = 0.0f
    private var level = Level.OFF
    private val pointPosition: PointF = PointF(0.0f, 0.0f)

    private lateinit var paint : Paint;

    private var _exampleString: String? = null // TODO: use a default from R.string...
    private var _exampleColor: Int = Color.RED // TODO: use a default from R.color...
    private var _exampleDimension: Float = 0f // TODO: use a default from R.dimen...

    private var textWidth: Float = 0f
    private var textHeight: Float = 0f

    /**
     * The text to draw
     */
    var exampleString: String?
        get() = _exampleString
        set(value) {
            _exampleString = value
            invalidateTextPaintAndMeasurements()
        }

    /**
     * The font color
     */
    var exampleColor: Int
        get() = _exampleColor
        set(value) {
            _exampleColor = value
            invalidateTextPaintAndMeasurements()
        }

    /**
     * In the example view, this dimension is the font size.
     */
    var exampleDimension: Float
        get() = _exampleDimension
        set(value) {
            _exampleDimension = value
            invalidateTextPaintAndMeasurements()
        }

    /**
     * In the example view, this drawable is drawn above the text.
     */
    var exampleDrawable: Drawable? = null

    private fun init(attrs: AttributeSet?, defStyle: Int) {
        // Load attributes
        val a = context.obtainStyledAttributes(
            attrs, R.styleable.CustomView, defStyle, 0
        )

        _exampleString = a.getString(
            R.styleable.CustomView_exampleString
        )
        _exampleColor = a.getColor(
            R.styleable.CustomView_exampleColor,
            exampleColor
        )
        // Use getDimensionPixelSize or getDimensionPixelOffset when dealing with
        // values that should fall on pixel boundaries.
        _exampleDimension = a.getDimension(
            R.styleable.CustomView_exampleDimension,
            exampleDimension
        )

        if (a.hasValue(R.styleable.CustomView_exampleDrawable)) {
            exampleDrawable = a.getDrawable(
                R.styleable.CustomView_exampleDrawable
            )
            exampleDrawable?.callback = this
        }

        a.recycle()

        isClickable = true;

        paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            style = Paint.Style.FILL
            textAlign = Paint.Align.CENTER
            textSize = 55.0f
            typeface = Typeface.create( "", Typeface.BOLD)
        }

        // Update TextPaint and text measurements from attributes
        invalidateTextPaintAndMeasurements()
    }

    override fun performClick(): Boolean {
        if (super.performClick()) return true

        level = level.next()
        contentDescription = resources.getString(level.label)

        invalidate()
        return true
    }

    private fun invalidateTextPaintAndMeasurements() {
        paint.let {
            it.textSize = exampleDimension
            it.color = exampleColor
            textWidth = it.measureText(exampleString)
            textHeight = it.fontMetrics.bottom
        }
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        // TODO: consider storing these as member variables to reduce
        // allocations per draw cycle.
        val paddingLeft = paddingLeft
        val paddingTop = paddingTop
        val paddingRight = paddingRight
        val paddingBottom = paddingBottom

        val contentWidth = width - paddingLeft - paddingRight
        val contentHeight = height - paddingTop - paddingBottom

        exampleString?.let {
            // Draw the text.
            canvas.drawText(
                it,
                paddingLeft + (contentWidth - textWidth) / 2,
                paddingTop + (contentHeight + textHeight) / 2,
                paint
            )
        }

        // Draw the example drawable on top of the text.
        exampleDrawable?.let {
            it.setBounds(
                paddingLeft, paddingTop,
                paddingLeft + contentWidth, paddingTop + contentHeight
            )
            it.draw(canvas)
        }

        paint.color = if (level == Level.OFF) Color.GRAY else Color.GREEN

        canvas.drawCircle((width / 2).toFloat(), (height / 2).toFloat(), radius, paint)

        val markerRadius = radius + RADIUS_OFFSET_INDICATOR
        pointPosition.computeXYForLevel(level, markerRadius)
        paint.color = Color.BLACK
        canvas.drawCircle(pointPosition.x, pointPosition.y, radius/12, paint)

        val labelRadius = radius + RADIUS_OFFSET_LABEL
        for (i in Level.values()) {
            pointPosition.computeXYForLevel(i, labelRadius)
            val label = resources.getString(i.label)
            canvas.drawText(label, pointPosition.x, pointPosition.y, paint)
        }
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        radius = (min(w, h) / 2.0 * 0.8).toFloat()
    }

    private fun PointF.computeXYForLevel(pos: Level, radius: Float) {
        // Angles are in radians.
        val startAngle = Math.PI * (9 / 8.0)
        val angle = startAngle + pos.ordinal * (Math.PI / 4)
        x = (radius * cos(angle)).toFloat() + width / 2
        y = (radius * sin(angle)).toFloat() + height / 2
    }
}